### PR TITLE
Allow specifying resources in Op/Asset params list

### PR DESCRIPTION
**Pull request by benpankow**
## Summary

Allows users to specify resource dependencies as parameters to an Asset or Op function.

Resource dependencies are identified by type annotations which either extend `ResourceDefinition` or are wrapped as `ResourceOutput[T]`. 

The first pattern is useful for "new"-style resources, which are Python classes that the user will interact with. In this case, the `ResourceDefinition` and the output from the resource function are identical. Here, a theoretical `S3Resource` implements `ResourceDefinition` and provides various interactions:

```python
@asset
def my_asset(s3: S3Resource):
    with s3.open_bucket(...):
        .... 
```

The second pattern is useful for function-style resources, which may output any arbitrary Python object. In this case, our `PostgresResource` outputs a `DatabaseConnection` object:

```python
@op
def my_op(postgres: ResourceOutput[DatabaseConnection]):
    postgres.query(...)
```

## Implementation

For the prototype, this approach was chosen over an approach which would use input names rather than annotations to identify which inputs are resources and which are other op/asset dependencies for a few reasons:

- Clarity: By using annotations, it is immediately clear to the reader which parameters are resources.
- Complexity: Asset dependencies are currently resolved immediately, when the decorator is applied. It would require a significant rewrite to delay this evaluation until the `Definitions` object containing the asset is evaluated.

## Test Plan

Unit tests

### CODE REVIEW

**Comment by benpankow** [comment]
Current dependencies on/for this PR:
* master
  * **PR #11322** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/11322" target="_blank">Graphite</a>  👈
    * **PR #11321** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/11321" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/11322?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/definitions/decorators/asset_decorator.py
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/decorators/asset_decorator.py:270]: I think we should strongly consider disallowing specifying `required_resource_keys`/`resource_defs` and the argument-based approach for now, similar to how we do not allow `config` argument and `config_schema`. We can always loosen this later, but I would prefer to be strict for now.

File: python_modules/dagster/dagster/_core/definitions/decorators/op_decorator.py
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/decorators/op_decorator.py:86]: same here

File: python_modules/dagster/dagster/_core/definitions/decorators/solid_decorator.py
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/decorators/solid_decorator.py:160]: same

**Review by schrockn — CHANGES_REQUESTED** [review]

**Review by schrockn — APPROVED** [review]
Looks great.

Given this is an exploration, I don't think we strictly need it, but make a note that we will need to test and support (or explicitly error) all different use cases where input names and upstream asset names might conflict with resource names.

**Review by schrockn — CHANGES_REQUESTED** [review]
Actually we definitely need some test cases with ops with inputs and assets that have upstream assets. Please add.

**Comment by benpankow** [comment]
Added some tests which go over upstream assets/ops. Doesn't handle the case in which names conflict, since we don't have a good way to resolve this at the moment without threading through resources to the execution planning logic (this may be what we want to do long-term).

File: python_modules/dagster/dagster/_core/definitions/decorators/solid_decorator.py
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/decorators/solid_decorator.py:165]: s/solid/op

**Review by schrockn — APPROVED** [review]
👍🏻

**Comment by benpankow** [comment]
**Graphite Merge Job**

Current status: 🔄 Waiting for checks to pass

This pull request is currently waiting for CI and mergeability checks to pass before being merged.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/11322?utm_source=stack-comment).


Job Reference: uHrT8DIrkchoQsyuEMKn

**Pull request merged**
